docs/YieldDistributor.md

```diff
@@ -0,0 +1,245 @@
+    participant YD as YieldDistributor
+    participant BREAD as BREAD Token
+    participant BB as BUTTERED_BREAD Token
+
+    User->>YD: getCurrentVotingPower(account)
+    activate YD
+    YD->>YD: Calculate period start and end
+    YD->>YD: getVotingPowerForPeriod(BREAD, start, end, account)
+    activate YD
+    YD->>BREAD: numCheckpoints(account)
+    alt numCheckpoints == 0
+        YD-->>YD: Return 0
+    else numCheckpoints > 0
+        YD->>BREAD: checkpoints(account, 0)
+        alt first checkpoint > end
+            YD-->>YD: Return 0
+        else first checkpoint <= end
+            loop Find latest checkpoint within interval
+                YD->>BREAD: checkpoints(account, index)
+            end
+            YD->>YD: Calculate initial voting power
+            loop Process remaining checkpoints
+                YD->>BREAD: checkpoints(account, index)
+                YD->>YD: Update total voting power
+                alt checkpoint <= start
+                    YD->>YD: Adjust for interval start
+                    YD-->>YD: Break loop
+                end
+            end
+        end
+    end
+    YD-->>YD: Return BREAD voting power
+    deactivate YD
+    
+    YD->>YD: getVotingPowerForPeriod(BUTTERED_BREAD, start, end, account)
+    activate YD
+    Note over YD: Same process as BREAD
+    YD-->>YD: Return BUTTERED_BREAD voting power
+    deactivate YD
+    
+    YD->>YD: Sum BREAD and BUTTERED_BREAD voting power
+    YD-->>User: Return total voting power
+    deactivate YD
+```
+Now, let's break down the getVotingPowerForPeriod function step by step:
+1. Input Validation:
+    - Check if the start time is before the end time.
+    - Ensure the end time is not after the current block.
+2. Initial Checkpoint Check:
+    - Get the total number of checkpoints for the account.
+    - If there are no checkpoints, return 0 voting power.
+3. Boundary Checks:
+    - If the first checkpoint is after the end of the interval, return 0 voting power.
+4. Find Relevant Checkpoints:
+    - Start from the latest checkpoint and move backwards.
+    - Find the most recent checkpoint that is within or before the end of the interval.
+5. Initialize Voting Power Calculation:
+    - Set the initial voting power based on the latest relevant checkpoint.
+    - Calculate the duration from this checkpoint to the end of the interval.
+6. Process Earlier Checkpoints:
+    - Iterate through earlier checkpoints, moving backwards in time.
+    - For each checkpoint:
+        - Calculate the voting power for the sub-interval between checkpoints.
+        - Add this to the total voting power.
+        - If the checkpoint is before or at the start of the interval:
+            - Adjust the voting power to exclude time before the interval start.
+            - Break the loop as we've covered the entire interval.
+7. Return Total Voting Power:
+    - The function returns the accumulated voting power over the specified interval.
+Key Points:
```

Comment:
my brain would be helped by having these key points precede the function breakdown so that i can go through the details with higher level context